src/StructuredLogger/BinaryLog.cs

```diff
@@ -133,7 +149,41 @@ public static Build ReadBuild(Stream stream, Progress progress, byte[] projectIm
 
             build.WaitForBackgroundTasks();
 
+            build.IsCompatibilityMode = eventSource.IsCompatibilityMode;
+            build.RecoverableReadingErrors = readerErrorsRegistry?.GetErrors().ToList() ??
+                new List<(ReaderErrorType errorType, BinaryLogRecordKind recordKind, int count)>();
+
             return build;
         }
+
+        private class ReaderErrorsRegistry

```

Comment:
Are you sure this level of sophistication is warranted? I wouldn't bother doing any grouping or anything. Just a flat list of simple messages/warnings/errors seems good enough. Or maybe they're not even necessary. All we really want to know is: did we read the whole thing successfully or was there any information we weren't able to handle?